griptape/loaders/csv_loader.py

```diff
@@ -13,27 +15,34 @@
 class CsvLoader(BaseLoader):
     embedding_driver: Optional[BaseEmbeddingDriver] = field(default=None, kw_only=True)
     delimiter: str = field(default=",", kw_only=True)
+    encoding: str = field(default="utf-8", kw_only=True)
 
-    def load(self, source: str, *args, **kwargs) -> list[CsvRowArtifact]:
-        return self._load_file(source)
-
-    def load_collection(self, sources: list[str], *args, **kwargs) -> dict[str, list[CsvRowArtifact]]:
-        return utils.execute_futures_dict(
-            {utils.str_to_hash(source): self.futures_executor.submit(self._load_file, source) for source in sources}
-        )
-
-    def _load_file(self, filename: str) -> list[CsvRowArtifact]:
+    def load(self, source: bytes | str, *args, **kwargs) -> ErrorArtifact | list[CsvRowArtifact]:
         artifacts = []
 
-        with open(filename, encoding="utf-8") as csv_file:
-            reader = csv.DictReader(csv_file, delimiter=self.delimiter)
-            chunks = [CsvRowArtifact(row) for row in reader]
+        if isinstance(source, bytes):
+            try:
+                source = source.decode(encoding=self.encoding)
+            except UnicodeDecodeError:
+                return ErrorArtifact(f"Failed to decode bytes to string using encoding: {self.encoding}")
+        elif not isinstance(source, str):
+            return ErrorArtifact(f"Expected source to be bytes or string, got {type(source)}")
 
-            if self.embedding_driver:
-                for chunk in chunks:
-                    chunk.generate_embedding(self.embedding_driver)
+        reader = csv.DictReader(StringIO(source), delimiter=self.delimiter)
+        chunks = [CsvRowArtifact(row) for row in reader]
 
+        if self.embedding_driver:
             for chunk in chunks:
-                artifacts.append(chunk)
+                chunk.generate_embedding(self.embedding_driver)
+
+        for chunk in chunks:
+            artifacts.append(chunk)
 
         return artifacts
+
+    def load_collection(
+        self, sources: Sequence[bytes | str], *args, **kwargs
+    ) -> dict[str, ErrorArtifact | list[CsvRowArtifact]]:
+        return cast(
+            dict[str, Union[ErrorArtifact, list[CsvRowArtifact]]], super().load_collection(sources, *args, **kwargs)
```

Comment:
Can we use pipe `|` syntax instead of `Union` here?

Reply:
pyright didn't like it and I couldn't quickly figure out the reason why it complains about using pipe in the casts. I don't think its worth the time to investigate.